Small Makefile Changes/Fixes

David-McKenna opened the pull request:
Hey Dave,

I came across a few oddities in the rawspec Makefile while trying to debug some other issues I was having, this PR contains 3 noticeable changes that fixed those oddities for me. Let me know if there's anyhting you want me to change or remove from the PR.

- The install target now requires the rawspec executable to be built before it will try to install it 
  - Previously, the install failed as the file was not found
- The CC variable will now respect environment variable, and only default back to GCC in the case that $CC is not set
  - Previously, this was overwriting any $CC environment variables, which caused issues for me as my default GCC install is too new for the Nvidia CUDA compiler
- Introduced a default value for CXX and swapped HOST_COMPILER to use it
  - NVCC needs a C++ capable compiler, or libcxx linked, in order to link CUDA code, gcc was raising errors here until g++ was used

Cheers,
David

david-macmahon commented:
This all looks good, but I wonder about the needing a C++ compiler part.  How did ever work before with just `gcc`?

David-McKenna commented:
I'm not too sure myself, maybe something changed in the CUDA runtime or GCC in recent versions? GCC can in theory change it's backend based on input file extensions or be forced to compile and link C++ code with the `-x` extension. But as of right now, neither locally with CUDA 10.2.89 or on a remote server with CUDA 9.2.148 will compile rawspec with just GCC for me, both installs raise the same linking errors, looking for C++ symbols.

Given the Makefile has a default path that refers to CUDA 8.0, which was released in 2012 and superseded in 2015, and only supported up to GCC-5, there's been a good chunk of time where the change could have been introduced on either the GCC or CUDA side.

david-macmahon commented:
Weird. I've built rawspec using gcc with CUDA versions 8.0, 10.0, 11.0.3, and 11.3.0 (and possibly others, but those are the ones I can confirm that I've built with).  Maybe it's a gcc version thing?  My version of gcc is fairly old (5.4.0).  Can you post the errors somewhere please?

David-McKenna commented:
I'm using newer versions of GCC (7, 8 and 10 on hand, apt indicates 5/6 are obsoleted so I can't easily grab them) so you're likely right there, I've switched back to the current master branch and changed the CC variable to accept the current env, testing on CUDA 10/11 locally all the versions raise the same linking issues (apart from cases where the GCC version isn't supported by the CUDA version).

[Here's a gist with the outputs of the different CUDA/GCC versions](https://gist.github.com/David-McKenna/af3258d99d7efc241bfd5053cf14432c), all giving roughly the same linking errors related to C++ symbols.

texadactyl commented:
@david-macmahon 

Here is a table of nvcc 10.2/gcc compatibility: https://docs.nvidia.com/cuda/archive/10.2/cuda-installation-guide-linux/index.html

Looking at the gists that you created:

* cuda10_gcc10.txt :: Clearly, cuda 10.2 and gcc 10 are not compatible.  If you are going to use cuda 10.2,  gcc/g++ must be version 8 or earlier.
* cuda10_gcc7.txt :: C++?  Rawspec is a pure C source repository.
* cuda10_gcc8.txt :: ditto
* cuda11_gcc8.txt :: ditto

I would suggest using the original rawspec Makefile with cuda 11 a,d gcc <= 8.

David-McKenna commented:
Hey @texadactyl,

This issue arose as our data processing docker images had the `gcc' alias bound to `gcc-10' due to some of the installed software needing C++20 support which wasn't available with gcc-9 or lower, hence we had the need to change the CC variable in rawspec to something else as the CUDA version on the nodes did not support gcc-10, as evidenced above.  

I believe that compatibility matrix is describing the GCC release version, I just did a bit of testing and found that NVCC seems to treat "gcc" and "gcc-N" differently when processing inputs to the `-ccbin' flag, and requires the g++ binary to be provided for any case other than a literal `gcc' input, even when both inputs are running the same executable[0], giving the behaviour of rawspec running fine as-is, but needing a C++ compiler if any changes are made. Nvidia even suggest that a C++ compiler is required in their docs of NVCC[0], where they say 

> A general purpose C++ host compiler is needed by nvcc in the following situations

And unfortunately, changing CC in the makefile to a C++ compatible compiler does not work due to differences in pointer philosophy between the languages raising errors[2]. As a result, in order to support a version of GCC that is not available through the specific `gcc' alias/link (and likely a similar situation for other nvcc-compatible compilers), a C++ compiler needs to be specified separately to nvcc, hence leading to the changes I submitted in this PR

[0] https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#supported-host-compilers
[1] https://gist.github.com/David-McKenna/ca12f4a5f2c34b9f291106b814c01f26
[2] https://gist.github.com/David-McKenna/8bf4986c2dd3e646bdf36d9d5a861700

texadactyl commented:
@David-McKenna 

Then, I suggest this in your scripts/docker files and leave the off-the-shelf Makefile as-is:
```
export HOST_COMPILER=gcc
make
```

David-McKenna commented:
Hey @texadactyl,

As I said, our `gcc` alias does not point to a `nvcc`-compatible version of GCC. While something along the lines of `HOST_COMPILER=g++-9 make` would work, why not make an explicit change so that other users don't have the opportunity to deal with this foot-gun, whereby in all non-standard cases the `-ccbin` input to `nvcc` needs to be a C++ compiler?

texadactyl commented:
From @David-McKenna in slack chat::

"Heyo, I think it's fair to split them up as nvcc is requesting a C++ compiler, even if all of the rawspec code is C-based, it just seems like it can lead to multiple headaches (such as when I was originally trying to track down what was happening here) if it isn't made explicitly clear.
I just checked and saw that it works under the assumption that you are using a gcc-based compiler, but I foresee see an issue where if I was using a different compiler for the CUDA compile, say clang or icc instead, that there would be binary incompatabilities between the C/C++ compiled code leading to further issues given that the CC variable can't be changed from gcc to anything else since the makefile doesn't allow it to be overwritten from the environment
These aren't the LOFAR/ASTRON machines, I'm using the Irish LOFAR station compute nodes I'm working with, separate to the Breakthrough ones installed on the site, I'm a bit locked down on what I can do/can't do in terms of software instllations, but my main limitng factor is the requirements of other users I'm supporting unfortunately, which is what got me stuck in this mess of compilers"

cc: @david-macmahon

texadactyl commented:
@david-macmahon 

I see David McKenna's point.  I agree with him in concept.  The .cu file is indeed a C++ source file by nvcc convention.

@David-McKenna

With one change to line 12 in the existing Makefile, I can get the .cu file compiled with g++ and the .c files compiled with gcc:
```
NVCC          := $(CUDA_PATH)/bin/nvcc -ccbin $(HOST_COMPILER)
```
to
```
NVCC          := $(CUDA_PATH)/bin/nvcc -ccbin g++
```
I have compiled all of rawspec with the new Makefile and regression tested against the baseline.

texadactyl commented:
@david-macmahon 

"How did it ever work before with just gcc?"

I think the answer is simply that there is no C++ syntax in any of the source and no C++ standard inclusions referenced.

The cuda include files are C/C++ neutral (E.g. #ifdef __cplusplus):
```
#include <cufft.h>
#include <cufftXt.h>
#include <helper_cuda.h>
```

David-McKenna commented:
> I think the answer is simply that there is no C++ syntax in any of the .cu source and no C++ standard inclusions referenced.

I think there's more to it than that. I still believe `nvcc` has an explicit edge case to detect `gcc` as the input, and find a `g++` binary for it, as I can add a fairly standard C++ 'hello world' snippet,

```#include <iostream>

int main() {
    std::cout << "Hello World";
    return 0;
}````

 to the end of `rawspec_gpu.cu`, and everything will still compile fine when `gcc` is passed to `ccbin`.
